### PR TITLE
Fix integer overflow in degenerate invoke_signed BPF syscalls

### PR DESCRIPTION
**Pull request by Mrmaxmeier**
#### Problem

Malicious BPF programs can trigger debug assertions in `check_instruction_size`.
`num_accounts` and `data_len` are read from guest memory.

#### Summary of Changes

- Use saturating_mul and add instead of overflowing arithmetic.
- Unrelated unsafe cleanup in `translate_slice_inner`

### CODE REVIEW

**Review by jackcmay — APPROVED** [review]

File: programs/bpf_loader/src/syscalls.rs [hunk 1]
```diff
@@ -348,7 +348,7 @@ fn translate_slice_inner<'a, T>(
     {
         Err(SyscallError::UnalignedPointer.into())
     } else if len == 0 {
-        Ok(unsafe { from_raw_parts_mut(0x1 as *mut T, len as usize) })
+        Ok(&mut [])
```
Comment (jackcmay) [inline programs/bpf_loader/src/syscalls.rs:351]: I suspect this change is breaking the stable-perf ci test:
https://buildkite.com/solana-labs/solana/builds/39465#641bf1cd-feda-4802-afd9-c8fcd6cbc302

Comment (jackcmay) [reply programs/bpf_loader/src/syscalls.rs:351]: nevermind, the issue in the size calc

File: programs/bpf_loader/src/syscalls.rs [hunk 2]
```diff
@@ -1471,7 +1471,7 @@ fn check_instruction_size(
     data_len: usize,
     invoke_context: &Ref<&mut dyn InvokeContext>,
 ) -> Result<(), EbpfError<BPFError>> {
-    let size = num_accounts * size_of::<AccountMeta>() + data_len;
+    let size = num_accounts.saturating_mul(size_of::<AccountMeta>().saturating_add(data_len));
```
Comment (jackcmay) [inline programs/bpf_loader/src/syscalls.rs:1474]: Order of operations:
```suggestion
    let size = num_accounts.saturating_mul(size_of::<AccountMeta>()).saturating_add(data_len);
```

Comment (Mrmaxmeier) [reply programs/bpf_loader/src/syscalls.rs]: Whoops, thanks! :slightly_smiling_face:

**Pull request merged**